### ISSUE TITLE
Lose entire library after drive disconnection

### ISSUE THREAD

**Issue by cactushydrocodone** [jellyfin/jellyfin#1714]
Hi,

Im currently not using jellyfin because of the android apps, but thats another point.
My question is, is this age old bug fixxed:
- Have medialibrary on a FUSE mountpoint
- Sometime you loose the mountpoint (Connection issue, Process reaper, unexpected circumstances, etc)
- Emby (bzw. jellyfin) deletes the entire library because it cant find the files anymore
- After a remount and restart of emby (bzw. jellyfin), the entire library is gone
- To restore, you need to do a full rescan, which can take days (My library is 60TB).

Workaround i had for this the problem:
- Keep a daily backup of the emby (bzw. jellyfin) data folder and restore it on mountpoint loss.

The original emby devs basically said: "ye we fix that, sometime."
That was 3 years ago.

Maybe i have luck with you guys.

**Comment by cvium** [comment]
It's not fixed yet, but we want to.

**Comment by EraYaN** [comment]
Kodi keeps the metadata in it's database for a while, and so re-adding the file at the same path will repopulate all data and watched status etc.

**Comment by joshuaboniface** [comment]
I've never actually experienced this myself, either on Emby or Jellyfin, despite my CephFS volume going away all the time. Is there some particular action that triggers this? Even if the files aren't there, Jellyfin shouldn't be randomly deleting the entries in the DB, unless I'm missing something?

**Comment by EraYaN** [comment]
I believe a scan can clean up “removed” files.

**Comment by RaymondSchnyder** [comment]
> 
> 
> I've never actually experienced this myself, either on Emby or Jellyfin, despite my CephFS volume going away all the time. Is there some particular action that triggers this? Even if the files aren't there, Jellyfin shouldn't be randomly deleting the entries in the DB, unless I'm missing something?

Its as easy as this:

1. Emby / Jellyfun runs in a container, the FUSE mount is added with a docker mount
2. When the FUSE Mount on the host "goes away", it also goes away in the container
3. Emby / Jellyfin cant find its files anymore, at the next rescan, it begins to delete everything it doesnt have anymore (I do use a regular scan every 4 hours)
4. Even if you dont have a regular scan activated, at the next restart of the container (emby/jellyfin), it deletes ALL METADATA, EVERYTHING. All knowledge it ever had about those files.

**Comment by RaymondSchnyder** [comment]
Another day on which ive lost a library.
The full rescan will take days, weeks.

Why did the emby devs did this like this?
It makes no fucking sense, im sorry.

**Comment by joshuaboniface** [comment]
I did find [this from their forums](https://emby.media/community/index.php?/topic/57153-prevent-emby-from-removing-missing-media/), which basically says that this is intentional behaviour to let Emby (and hence JF) be "self maintaining". And I see their point - if you delete a file, then rescan, the intent is that that file will be detected as missing and be removed from the library. So that does make sense in a way. That said, I definitely think it needs to handle the "whole library is missing" situation a little better. From their talk there it seems like it's not a simple problem, and...

>Why did the emby devs did this like this?

Yea, that comes up a lot in our discussions...

As a quick workaround, my first though would be to set some sort of heath check in the Docker container that checks if the mount is available (`mount | grep /mountpoint`, or `test -f /mountpoint/testfile_that_is_never_deleted`, or something to that effect) and kill the container if the healthcheck fails. Or disable auto-scanning, and scan on some sort of event in the filesystem (i.e. use `inotifywait`) via the API. Neither is a particularly nice workaround in Docker I don't think; I depend Jellyfin on my CephFS mount unit but that's at a VM level.

**Comment by JustAMan** [comment]
Another solution from JF point of view might be adding a checkbox which effectively disables removing files from a database even if they're gone, like if you never intend to delete a file, plus making a library rescan which is run manually to drop the files.

So they never get lost by themselves but the administrator can remove them and force a re-check when he/she is sure the library is present.

**Comment by RaymondSchnyder** [comment]
> I (well, @anthonylavado) did find [this from their forums](https://emby.media/community/index.php?/topic/57153-prevent-emby-from-removing-missing-media/), which basically says that this is intentional behaviour to let Emby (and hence JF) be "self maintaining". And I see their point - if you delete a file, then rescan, the intent is that that file will be detected as missing and be removed from the library. So that does make sense in a way. That said, I definitely think it needs to handle the "whole library is missing" situation a little better. From their talk there it seems like it's not a simple problem, and...
> 
> > Why did the emby devs did this like this?
> 
> Yea, that comes up a lot in our discussions...
> 
> This discussion was occurring around 3.2, so any changes they made should be in Jellyfin's codebase, unless it was later remove. They're never detailed enough in their comments there ("check latest beta" and similar shit, rather than "see this PR for the fix"...) to really know where or when it was fixed without serious digging, but it should be date-searchable if someone really does want to dig deep into it.
> 
> As a quick workaround, my first though would be to set some sort of heath check in the Docker container that checks if the mount is available (`mount | grep /mountpoint`, or `test -f /mountpoint/testfile_that_is_never_deleted`, or something to that effect) and kill the container if the healthcheck fails. Or disable auto-scanning, and scan on some sort of event in the filesystem (i.e. use `inotifywait`) via the API. Neither is a particularly nice workaround in Docker I don't think - I depend Jellyfin on my CephFS mount unit but that's at a VM level - but should definitely prevent the full failure you're getting.
> 
> Another option might be to create a "library" filled with symlinks to the FUSE filesystem, and then scan _that_ "library", so that technically the files never "go away", which is one heck of a hack but should also work if my understanding of JF's file scanning is correct.

T

> I (well, @anthonylavado) did find [this from their forums](https://emby.media/community/index.php?/topic/57153-prevent-emby-from-removing-missing-media/), which basically says that this is intentional behaviour to let Emby (and hence JF) be "self maintaining". And I see their point - if you delete a file, then rescan, the intent is that that file will be detected as missing and be removed from the library. So that does make sense in a way. That said, I definitely think it needs to handle the "whole library is missing" situation a little better. From their talk there it seems like it's not a simple problem, and...
> 
> > Why did the emby devs did this like this?
> 
> Yea, that comes up a lot in our discussions...
> 
> This discussion was occurring around 3.2, so any changes they made should be in Jellyfin's codebase, unless it was later remove. They're never detailed enough in their comments there ("check latest beta" and similar shit, rather than "see this PR for the fix"...) to really know where or when it was fixed without serious digging, but it should be date-searchable if someone really does want to dig deep into it.
> 
> As a quick workaround, my first though would be to set some sort of heath check in the Docker container that checks if the mount is available (`mount | grep /mountpoint`, or `test -f /mountpoint/testfile_that_is_never_deleted`, or something to that effect) and kill the container if the healthcheck fails. Or disable auto-scanning, and scan on some sort of event in the filesystem (i.e. use `inotifywait`) via the API. Neither is a particularly nice workaround in Docker I don't think - I depend Jellyfin on my CephFS mount unit but that's at a VM level - but should definitely prevent the full failure you're getting.
> 
> Another option might be to create a "library" filled with symlinks to the FUSE filesystem, and then scan _that_ "library", so that technically the files never "go away", which is one heck of a hack but should also work if my understanding of JF's file scanning is correct.

This is the best answer ive got on that topic till yet. 
Ive made some topics on the emby forum, but got my issue never resolved / never got any new ideas like the ones you brought to the table right now.

I tried jellyfin a couple days ago, its really neat and im looking to replace my proprietary emby with jellyfin as soon as possible.

**Comment by joshuaboniface** [comment]
Thanks, hopefully at least one of those workarounds helps :-)

**Comment by EraYaN** [comment]
Well at some point we could adopt the "Kodi" way, it can detect a file missing and a full mount point / station missing, and handles these differently when cleaning the database. (It asks when the whole mount is missing otherwise it just cleans em up)

**Comment by hereisderek** [comment]
please have a look into this, this is much needed

some potential measurement includes:

- allows the user to manually opt in to `manual clean up only`
- allows the user to select the auto clean schedule (weekly monthly)
- only clean up the metadata that's expired for a specified time

**Comment by luispabon** [comment]
Kodi fixes this problem by only adding new found media when rescanning, and having a library cleanup action that removes missing files as a separate action. The library cleanup has to be triggered manually.

**Comment by PrplHaz4** [comment]
> Kodi fixes this problem by only adding new found media when rescanning, and having a library cleanup action that removes missing files as a separate action. The library cleanup has to be triggered manually.

Sounds like a reasonable approach - but probably requires the ability to handle missing media gracefully on all clients...

**Comment by MattMckenzy** [comment]
I've had this issue at least once a month since I've started using Jellyfin, it's so very frustrating to spend hours recreating all my libraries. I'm about to fork Jellyfin and try to fix this myself I think.

Thought of a good solution: create a hidden file at root of library upon creation. Check for file when scanning and if not found, don't clean up library but log an error and skip scanning it.

**Comment by ArabCoders** [comment]
While i think this problem should be fixed in jellyfin, as workaround, you could have a cronjob that check for mount status, if it's not available "transport disconnected" or w/e, you could kill off jellyfin instance until the mount become available again, it's a batch work job but at least you dont have to rescan the entire library specially when it's in the cloud and mounted via rclone or whatever. I think there is a related config option for systemd that kill off services if mount is disconnected https://blog.thewatertower.org/2019/12/29/mount-dependencies-with-systemd/

**Comment by joshuaboniface** [comment]
> I'm about to fork Jellyfin and try to fix this myself I think.

Well, we'd absolutely welcome a PR to fix this, since no one so far has had the time or inclination to fix it.

**Comment by MattMckenzy** [comment]
I think I might have gotten something working for this. I've created a draft pull-request at #8814 in case someone would like to have a look or try it out!

I still want to build it as a container and deploy it to my home set-up for testing before I sign off on the pull request, hopefully it won't take too long.

**Comment by AndrewDavis** [comment]
This is a simple matter of needing graceful degredation. If a drive is no longer there, then simply hide everything in the drive metadata. Once the drive is there again, show everything again. Absolutely no other action should be taken by Jellyfin.

Why is this different than any cleanup or refresh process? Simple: It's a matter of whether the drive itself is accessible. If the drive is accessible, but a file is no longer there, that's a different matter. Easy logic to separate the two.

Simply put a boolean in the drive's metadata or whatever, and then add a check to see whether the drive is there. Done.

**Comment by Skalmansbror** [comment]
I agree. This is an absolute disgrace. 
I have a collection of ~2000 digitally recorded old VHS movies. 
The task of getting posters and backdrops for them all was a REAL hassle. Weeks and weeks of work. 
I finally got it all set up, and I was thrilled to be able to browse my gallery digitally. 

My NAS then lagged a bit after rebooting, jellyfin recognized no files were present, and wiped my entire library. 
To go through it all again feels absolutely disheartening, especially since it can happen again at any moment!

Fix this issue asap! It sucks!

**Comment by Bond-009** [comment]
This is being worked on: https://github.com/jellyfin/jellyfin/pull/9291

@AndrewDavis If it's so simple you can always contribute a fix.

@Skalmansbror you can put the posters and backdrops in the same folder as your media, that way they will never get deleted by Jellyfin.

**Comment by AndrewDavis** [comment]
> @AndrewDavis If it's so simple you can always contribute a fix.

No thanks, I'm just here to tell you how to do your job

**Comment by Dryadwoods** [comment]
> version 10.8.9

Well... I am also using **version 10.8.9**
And sadly I can say that this is **not fixed.**

This week I was busy with life and I only noticed 2 or 3 days later that I had problems with my NAS.
Meanwhile, jellyfin during his automatic tasks did something that it should...... since was NOT able to reach the files anymore... therefor ALL my libraries were lost..... really everything.

I am already refreshing them and this is going to take me forever to finish.

**Comment by Fruckiwacki** [comment]
While I am also interested in a solution, it just baffles me that people speak about their precious and hard to curate libraries, yet seemingly have no backup solution in place. 

If whatever catastrophic damage happens to your library, simply restore it from a backup.

**Comment by Dryadwoods** [comment]
> While I am also interested in a solution, it just baffles me that people speak about their precious and hard to curate libraries, yet seemingly have no backup solution in place.
> 
> If whatever catastrophic damage happens to your library, simply restore it from a backup.

When I say that the "library" was lost, I meant the Jellyfin database data... not the files it self.
Of course, that also some of the "mappings" or CUSTOM metadatas (might be in a file, or in the DB) might also be affected.
And when you refer backup... well... what type are you referring too that applies to this case?

**Comment by luispabon** [comment]
I back up everything, more so irreplaceable data like that.

**Comment by Fruckiwacki** [comment]
> > While I am also interested in a solution, it just baffles me that people speak about their precious and hard to curate libraries, yet seemingly have no backup solution in place.
> > If whatever catastrophic damage happens to your library, simply restore it from a backup.
> 
> When I say that the "library" was lost, I meant the Jellyfin database data... not the files it self (movies, subtitles). Of course, that also some of the "mappings" or CUSTOM metadatas (might be in a file, or in the DB) might also be affected. And when you refer backup... well... what type are you referring too that applies to this case?

I refer to the data you mentioned - the Jellyfin config directory, which includes all metadata, configuration etc.

**Comment by Dryadwoods** [comment]
I completely agree with the use of backups, no discussion on that.

Although, I might highlight that fact that not everyone knows how to perform backups (of different sources or types) and that is really not the point of this open ticket, the issue is that Jellyfin is losing it, and that should be the main discussion.

I would like to suggest that **one easy way to fix this problem** would be to have a **BIT field** associated with one library that says "**Lock library**"... in the sense that no matter what happens, simple do not change anything about it. Keep the database entries, and everything else the way it is, do not delete, neither do not add or update it.

**Comment by cvium** [comment]
Feel free to take a look through the code base. You might feel different.

**Comment by AndrewDavis** [comment]
> While I am also interested in a solution, it just baffles me that people speak about their precious and hard to curate libraries, yet seemingly have no backup solution in place.

Haha, if Jellyfin were actually causing all the source media files to be deleted, that would truly make it a trash product. Fortunately that's not what's happening here, just the metadata.

---

Are we sure we're talking about the same issue here? I.e. when your metadata is lost, are you able to get it all back by simply rescanning everything? I am; of course; it takes a while, which is the obnoxious part. The original problem mentioned which I also have is that the drive is disconnected when Jellyfin starts, so it loses the metadata.

---

> I would like to suggest that one easy way to fix this problem would be to have a BIT field associated with one library that says "Lock library"... in the sense that no matter what happens, simple do not change anything about it. Keep the database entries, and everything else the way it is, do not delete, neither do not add or update it.

No, this would not work, because 1. you'll still lose the metadata if you unlock it when the drive is disconnected, 2. you'll lose the auto-update feature, and 3. you'll have to unlock and relock it every time you add or remove anything. The only proper way to fix this problem would be to do what I said [here](https://github.com/jellyfin/jellyfin/issues/1714#issuecomment-1442855762), it's just a matter of implementation, how would that be coded. If the drive cannot even be accessed, then no metadata is updated/refreshed/modified for said drive, it is only hidden from the user, and assumed to still be there once accessible again.

**Comment by AndrewDavis** [comment]
Okay, that makes sense, although I think that's likely a non-standard case. Glad you're going to save the planet by keeping a server offline...

There's a problem with your request though in that obviously, in what would probably be a more typical case of a drive **not** turning on when offline files are requested, Jellyfin would then be presenting the user with files which are not available should they request them. Maybe that's desirable but it should at least signal to the user somehow that such is the case.

**Comment by sannidhyaroy** [comment]
Any updates on the progress?

**Comment by cactushydrocodone** [comment]
As its a complicate thing to fix, i would suggest documentating an official, stable workaround

**Comment by sannidhyaroy** [comment]
> As its a complicate thing to fix, i would suggest documentating an official, stable workaround

Do you know how to properly keep a Metadata backup? Can I just set up a daily cronjob that copies the entire metadata folder and encrypts with my public key? Since the metadata is stored as hashes as far as I know, can I restore from my backup in case stuffs get deleted when Jellyfin can't find my media?

**Comment by AndrewDavis** [comment]
What's been working well for me is simply turning off the auto refresh stuff. When I add more content I go and trigger a manual refresh, it's not hard and it's very quick even with a big library. No disconnect or similar problems.

I recommend everyone do this for now.

**Comment by cactushydrocodone** [comment]
@sannidhyaroy 
I would cache the filesystem locally, only the structure, so filenames, paths, etc.
That way jellyfin, even when the remote backend disappears, still thinks the files are there, but it cant really access it. That would cause a "your device cant play this file now" or whatever error but jellyfin wouldnt delete the file, as its there, just not valid.

**Comment by AndrewDavis** [comment]
> I would cache the filesystem locally
There's no such thing. Either you copy the files, which defeats a primary purpose of using external storage - multimedia takes a LOT of disk space - or you make symbolic links or similar, which will still fail you. If you had a media server than copied all the files off of a device everyone would become irate; bad idea.

**Comment by sannidhyaroy** [comment]
> I would cache the filesystem locally, only the structure, so filenames, paths, etc.

@cactushydrocodone How do you do that?

**Comment by ArabCoders** [comment]
It is possible to cache the file metadata via rclone, not sure if they survive long enough if you don't notice the drives are gone tho

**Comment by AndrewDavis** [comment]
Don't downvote me, I told you the truth. If you don't like it or understand it then that's your problem. If you're going to try to suggest solutions then you should first actually understand what you're talking about.

If you want to clone stuff then do that manually, that's not Jellyfin's prerogative and it's not a Jellyfin-solution to this problem.

**Comment by hereisderek** [comment]
> @sannidhyaroy I would cache the filesystem locally, only the structure, so filenames, paths, etc. That way jellyfin, even when the remote backend disappears, still thinks the files are there, but it cant really access it. That would cause a "your device cant play this file now" or whatever error but jellyfin wouldnt delete the file, as its there, just not valid.

I'd love to know a way of doing that too, it'd solve a lot of other problems i'm having.

**Comment by centralware** [comment]
Has anyone made any progress (or potentials)?
Today's my first day of laying eyes on JF (online - I haven't installed) so I decided to go through my list of gripes we had back in the day of Twonky, and offline libraries was high on that list.  (NOT referring to JF's "Download->Offline" client-side.)

Logic while starting a scan on the server:
[TEST] Get a directory structure of the library being scanned
* IF EMPTY, set OFFLINE = True and **ignore** all existing content in the database
Skip the remainder of this library and move onto the next, set OFFLINE = False afterward
In the web UI, give the admin the option to manually set OFFLINE for each library
* When manually set, skip scanning the library, leave database items alone
* Only when a library is listed as ONLINE should old file cleanup commence

All of the above assumes the server is keeping a cached copy of library graphics to be used if/when needed by the client(s).
Cached graphics are purged if the library being scanned is ONLINE and assumes file sizes/dates are different or missing - or maybe also manually in the web UI on a per-library basis in case someone needed to manually wipe out the cache for a complete library.

If the above can be accomplished with reasonable effort and time, unavailable resources due to networking and the likes should act as though everything is fine until the client goes to access a file that's offline, and they're given the less-than-descriptive "unavailable" error everyone loves so much; but syncing with the server should be seamless.  Additionally, it allows people (such as myself) to intentionally take libraries saved on USB, NAS...  heck, FLOPPY or TAPE DRIVE if there's an itch...  and manually set each library as being "offline" which in my case, let's say I wanted to have wedding photos/videos implemented on the server.  It's far from being an every day usage item, so having them able to be plugged in and "re-enabled" would be the way to go.  (A concept from back in the BBS days.)

In the BBS days, we had the server show OFFLINE files as a greyed out item; when clicked you'd get file specs and a storage ID (so it made finding the file a breeze.)  This would require client-side updates and doesn't entail anything to do with fixing the problem at hand, so it's a "nice to have" more than anything. :)  With that, it may not hurt to add (to the nice-to-have) that when an unavailable item is clicked, a separate log (and web UI?) is generated telling Admin "Hey, such-and-such USER wanted to play \PATH\TO\FILENAME.EXT from library ABC123..."

**Comment by ShrirajHegde** [comment]
> It should have some tags automatically assigned to that particular media which is not available rather than removing and scanning from zero. My library was deleted a few times as NAS is on other system and Proxmox is running Jellyfin in LXC Container. I tried to delay the startups but sometimes it just doesn't work.
> 
> I went with the Symbolic Links solution. Links will remain there and a cron job runs the command to copy the newer links.
> 
> `cp -aR -s /mnt/movies/ /home`
> 
> and for cron job
> 
> `0 * * * * cp -aR -s /mnt/movies/ /home > /dev/null 2>&1`
> 
> It creates the exact replica with Symbolic Links also you will have to update the Library Paths in Jellyfin
> 
> `/home/movies/` is the new path in this scenario

How does cleanup work in case you remove something?

**Comment by lukefor** [comment]
> How does cleanup work in case you remove something?

I ended up doing something similar and resolved cleanup of deleted items with
```
find . -xtype l -delete
find . -type d -empty -delete
```
Handling renames is a little more involved, you can either scrap the symlinks and recreate them all (not ideal) or have some way of tracking and deleting outdated but valid symlinks, such as touching the symlinks and then tracking their mtime

**Comment by Xitee1** [comment]
For me the symlink workaround sadly doesn't work. Jellyfin hasn't wiped the library but I'm completely unable to play anything. It always gives me an error when I try to play something.
I'm using the linuxserver docker image and specified the symlinked version of my library in the docker mount.

**Comment by nichobi** [comment]
> For me the symlink workaround sadly doesn't work. Jellyfin hasn't wiped the library but I'm completely unable to play anything. It always gives me an error when I try to play something. I'm using the linuxserver docker image and specified the symlinked version of my library in the docker mount.

The issue here would be that the symlinks aren't resolvable within the docker container. You'd need to also mount the actual directory containing the files, at the exact same path the symlinks point to. See [https://stackoverflow.com/questions/38485607/](https://stackoverflow.com/questions/38485607/)

**Comment by dpieski** [comment]
I know it wouldn't solve the underlying problem, but couldn't you just remove the trigger for the "Scan Media Library" task and only run it when you know the media is attached? It would still "clean up" files not found when you ran it, but it wouldn't just run without you telling it to?

**Comment by crobibero** [comment]
This was fixed with 10.9 where Jellyfin will no longer purge data if an entire library is empty.

**Comment by Xitee1** [comment]
> I just had this happen so it does not appear to be fixed. Running version 10.9.11

I've tested it last weekend and it appeared to work. Removed the volume in docker and started it. All media was there. It wasn't perfect because there was no note about missing media and when trying to play something, the watched state was reset. After adding the volume with the media again, everything was still there. I'm using the linuxserver image.
